components/raftstore/src/store/fsm/peer.rs

```diff
@@ -987,36 +988,98 @@ where
+    // first be checked and if neither is found, then a Raft Admin Command will be
+    // proposed to persist the state in the RegionLocalState to lock the region.
+    // Once this Command is submitted, a callback will be invoked to call msg
+    // PrepareFlashback again to wait for the new change to be applied completely.
+    // So we can perform the flashback safely later. We make FlashbackToVersion
+    // become a two-phase request, and lock the region as the first phase.
     fn on_prepare_flashback(&mut self, ch: Sender<bool>) {
+        let region_id = self.region().get_id();
         info!(
             "prepare flashback";
-            "region_id" => self.region().get_id(),
+            "region_id" => region_id,
             "peer_id" => self.fsm.peer.peer_id(),
         );
-        if self.fsm.peer.flashback_state.is_some() {
-            ch.send(false).unwrap();
-            return;
+        // check persist state
+        let state_key = keys::region_state_key(region_id);
+        if let Some(state) = self
+            .ctx
+            .engines
+            .kv
+            .get_msg_cf::<RegionLocalState>(CF_RAFT, &state_key)
+            .unwrap()
+        {
+            info!(
+                "Prepare Flashback, at now flashback persist state is: {} for {:?} ",
+                state.get_region().get_is_in_flashback(),
+                self.region(),
+            );
+            if !state.get_region().get_is_in_flashback() {
+                // callback itself to wait for the new change to be applied completely.
+                let raft_router_clone = self.ctx.router.clone();
+                let ch_clone = ch.clone();
+                let cb = Callback::write(Box::new(move |resp| {
+                    if resp.response.get_header().has_error() {
+                        ch.send(false).unwrap();
+                        error!("send flashback prepare msg failed"; "region_id" => region_id);
+                        return;
+                    }
+                    raft_router_clone
+                        .force_send(
+                            region_id,
+                            PeerMsg::SignificantMsg(SignificantMsg::PrepareFlashback(ch_clone)),
+                        )
+                        .unwrap();
+                }));
+                // set persist state by a Raft Admin Command
+                let req = {
+                    let mut request = new_admin_request(region_id, self.fsm.peer.peer.clone());
+                    let mut admin = AdminRequest::default();
+                    admin.set_cmd_type(AdminCmdType::PrepareFlashback);
+                    request.set_admin_request(admin);
+                    request
+                };
+                self.propose_raft_command(req, cb, DiskFullOpt::AllowedOnAlmostFull);
+                return;
+            }
+            // check memory state
+            if self.fsm.peer.flashback_state.is_none() {
```

Comment:
Better cover the case that `self.fsm.peer.flashback_state` is not `None` but the inside `ch` is.